src/python/pants/util/strutil.py

```diff
@@ -335,3 +342,38 @@ def wrapper(func: Callable[P, R]) -> Callable[P, R]:
         return func
 
     return wrapper
+
+
+class _JsonEncoder(json.JSONEncoder):
+    """Allow us to serialize everything, with a fallback on `str()` in case of any esoteric
+    types."""
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if dataclasses.is_dataclass(o):
+            return dataclasses.asdict(o)
+        if isinstance(o, abc.Mapping):
+            return dict(o)
+        if isinstance(o, (abc.Sequence, OrderedSet, FrozenOrderedSet)):
+            return list(o)
+        if isinstance(o, (Digest,)):
+            return {
+                "fingerprint": o.fingerprint,
+                "serialized_bytes_length": o.serialized_bytes_length,
+            }
```

Comment:
Hmm.. that's a good point. Maybe the class name for dataclasses should be part of the encoded value to distinguish otherwise similar instances that share the same set of fields (and values) but for different types. Not that I think that will be an issue more than in theory but..

Also, for the Digest, why not use just the fingerprint? (possibly with something indicating it is a digest fingerprint) as the length information is already embedded in the fingerprint, right?

Reply:
No, the length isn't embedded in the fingerprint. Thats why they both exist (AFAIK)

Reply:
I mean, the fingerprint is based on the content right? so if you change the length, the content change, so the fingerprint is a property also of the content length.

Reply:
I'm guessing the length is there mostly for our sake to give an idea of how large a digest is..

Reply:
Yea, it's not necessary to include it when serializing unless you hope to deserialize later. It's included because it's used for various optimizations in the storage backend.

Reply:
Ah. I'll edit